extensions/amp-ad/0.1/amp-ad-image.js

```diff
@@ -0,0 +1,101 @@
+/**
+ * Copyright 2016 David Karlin, Bachtrack Ltd. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {isLayoutSizeDefined} from '../../../src/layout';
+import {getLifecycleReporter} from '../../../ads/google/a4a/performance';
+import {user} from '../../../src/log';
+import {getBatchManager} from './amp-ad-batch-manager.js';
+import {templatesFor} from '../../../src/template';
+
+/** @const {!string} Tag name for 3P AD implementation. */
+export const TAG_AD_IMAGE = 'amp-ad-image';
+
+export class AmpAdImage extends AMP.BaseElement {
+
+  /** @param {!AmpElement} element */
+  constructor(element) {
+    super(element);
+
+    /** @private {string} The base URL of the ad server for this ad */
+    this.url_ = element.getAttribute('data-url');
+
+    /** @private {boolean} Whether this is the batch master */
+    this.isBatchMaster_ = false;
+
+    /** @private {string} A string identifying this ad slot: the server's responses will be keyed by slot */
+    this.slot_ = element.getAttribute('data-slot');
+    user().assert(!this.slot_ || this.slot_.match(/^[0-9a-z]+$/),
+        'imagead slot should be alphanumeric: ' + this.slot_);
+
+    // Ensure that there are templates in this ad
+    const templates = element.querySelectorAll('template');
+    user().assert(templates.length > 0, 'Missing template in imagead');
```

Comment:
Should these be handled in `#buildCallback`? I'm not sure if we can handle a throw in the constructor.